Incorrect "duplicate key" warnings when using `#line` as key in dictionary literals

**Describe the bug**
The latest trunk toolchain (incorrectly) reports a duplicate dictionary key warning when using `#line` on different lines (a pattern [commonly used by `swift-syntax` tests](https://github.com/apple/swift-syntax/blob/114e16a7fc057a3aa85cff4ad1f4be8e143db240/Tests/SwiftSyntaxBuilderTest/IntegerLiteralTests.swift#L7-L16)). For example:

```swift
let x = [
  #line: "a",
  #line: "b",
]
print(x)
```

compiled with

```
Apple Swift version 5.8-dev (LLVM 03bd949eeee4761, Swift 2c40a850df6f409)
Target: arm64-apple-macosx12.0
```

produces

```swift
dict-line.swift:2:3: warning: dictionary literal of type '[Int : String]' has duplicate entries for #line literal key
  #line: "a",
  ^
dict-line.swift:2:3: note: duplicate key declared here
  #line: "a",
  ^~~~~~~~~~~
            
dict-line.swift:3:3: note: duplicate key declared here
  #line: "b",
  ^~~~~~~~~~~
            
[3: "b", 2: "a"]
```

**Steps To Reproduce**
1. Compile the program above with `Apple Swift version 5.8-dev (LLVM 03bd949eeee4761, Swift 2c40a850df6f409)` (August 29, 2022 snapshot)

**Expected behavior**

The same behavior as on Swift 5.6:

```
swift-driver version: 1.45.2 Apple Swift version 5.6.1 (swiftlang-5.6.0.323.66 clang-1316.0.20.12)
Target: arm64-apple-macosx12.0
```

```swift
[3: "b", 2: "a"]
```

**Environment (please fill out the following information)**
 - OS: macOS 12.5.1
 - Xcode Version/Tag/Branch: Apple Swift version 5.8-dev (LLVM 03bd949eeee4761, Swift 2c40a850df6f409)

--- a/lib/Sema/MiscDiagnostics.cpp
+++ b/lib/Sema/MiscDiagnostics.cpp
@@ -5038,13 +5038,44 @@ diagnoseDictionaryLiteralDuplicateKeyEntries(const Expr *E,
 
   private:
     std::string getKeyStringValue(const LiteralExpr *keyExpr) {
-      if (isa<MagicIdentifierLiteralExpr>(keyExpr)) {
-        return keyExpr->getLiteralKindDescription().str();
+      if (auto *MLE = dyn_cast<MagicIdentifierLiteralExpr>(keyExpr)) {
+        return getMagicLiteralKeyValue(MLE);
       }
       std::string out;
       llvm::raw_string_ostream OS(out);
       keyExpr->printConstExprValue(&OS, /*additionalCheck=*/nullptr);
       return out;
+    }
+
+    std::string getMagicLiteralKeyValue(const MagicIdentifierLiteralExpr *MLE) {
+      auto magicLiteralValue = MLE->getLiteralKindDescription().str();
+      switch (MLE->getKind()) {
+      case MagicIdentifierLiteralExpr::DSOHandle:
+      case MagicIdentifierLiteralExpr::FileID:
+      case MagicIdentifierLiteralExpr::FileIDSpelledAsFile:
+      case MagicIdentifierLiteralExpr::FilePath:
+      case MagicIdentifierLiteralExpr::FilePathSpelledAsFile:
+      case MagicIdentifierLiteralExpr::Function:
+        break;
+      // Those are literals that can evaluate to different values in a
+      // dictionary literal declaration context based on source position
+      // so we need to consider that position as part of the literal value.
+      case MagicIdentifierLiteralExpr::Column: {
+        unsigned int column;
+        std::tie(std::ignore, column) =
+            Ctx.SourceMgr.getPresumedLineAndColumnForLoc(MLE->getStartLoc());
+        magicLiteralValue += ":" + std::to_string(column);
+        break;
+      }
+      case MagicIdentifierLiteralExpr::Line: {
+        unsigned int line;
+        std::tie(line, std::ignore) =
+            Ctx.SourceMgr.getPresumedLineAndColumnForLoc(MLE->getStartLoc());
+        magicLiteralValue += ":" + std::to_string(line);
+        break;
+      }
+      }
+      return magicLiteralValue;
     }
 
     std::string getKeyStringValueForDiagnostic(const LiteralExpr *keyExpr) {

--- a/test/Sema/diag_dictionary_keys_duplicated.swift
+++ b/test/Sema/diag_dictionary_keys_duplicated.swift
@@ -209,3 +209,21 @@ let _: [String: String] = [
   "\(a)": "B",
   "\(1)": "C"
 ]
+
+// https://github.com/apple/swift/issues/60873
+let _: [Int: String] = [
+  #line: "A",
+  #line: "B"
+]
+
+let _: [Int: String] = [#line: "A", #line: "B"] // expected-warning{{dictionary literal of type '[Int : String]' has duplicate entries for #line literal key}}
+// expected-note@-1{{duplicate key declared here}} {{25-35=}} {{35-36=}}
+// expected-note@-2{{duplicate key declared here}} {{37-47=}} {{35-36=}}
+
+let _: [Int: String] = [#column: "A", #column: "B"] // OK
+
+let _: [Int: String] = [
+  // expected-note@+1{{duplicate key declared here}} {{3-15=}} {{15-16=}} 
+  #column: "A", // expected-warning{{dictionary literal of type '[Int : String]' has duplicate entries for #column literal key}}
+  #column: "B"  // expected-note{{duplicate key declared here}} {{3-16=}} {{227:15-16=}}
+]

